src/linker/Linker.Steps/ProcessLinkerXmlStepBase.cs

```diff
@@ -0,0 +1,491 @@
+
+		protected virtual void ProcessTypes (AssemblyDefinition assembly, XPathNodeIterator iterator, bool warnOnUnresolvedTypes)
+		{
+			iterator = iterator.Current.SelectChildren (TypeElementName, XmlNamespace);
+			while (iterator.MoveNext ()) {
+				XPathNavigator nav = iterator.Current;
+
+				if (!ShouldProcessElement (nav))
+					continue;
+
+				string fullname = GetFullName (nav);
+
+				if (fullname.IndexOf ("*") != -1) {
+					if (ProcessTypePattern (fullname, assembly, nav))
+						continue;
+				}
+
+				TypeDefinition type = assembly.MainModule.GetType (fullname);
+
+				if (type == null && assembly.MainModule.HasExportedTypes) {
+					foreach (var exported in assembly.MainModule.ExportedTypes) {
+						if (fullname == exported.FullName) {
+							var resolvedExternal = ProcessExportedType (exported, assembly);
+							if (resolvedExternal != null) {
+								type = resolvedExternal;
+								break;
+							}
+						}
+					}
+				}
+
+				if (type == null) {
+					if (warnOnUnresolvedTypes)
+						Context.LogWarning ($"Could not resolve type '{fullname}' specified in {_xmlDocumentLocation}", 2008, _xmlDocumentLocation);
+					continue;
+				}
+
+				ProcessType (type, nav);
+			}
+		}
+
+		protected virtual TypeDefinition ProcessExportedType (ExportedType exported, AssemblyDefinition assembly) => exported.Resolve ();
+
+		void MatchType (TypeDefinition type, Regex regex, XPathNavigator nav)
+		{
+			if (regex.Match (type.FullName).Success)
+				ProcessType (type, nav);
+
+			if (!type.HasNestedTypes)
+				return;
+
+			foreach (var nt in type.NestedTypes)
+				MatchType (nt, regex, nav);
+		}
+
+		protected virtual bool ProcessTypePattern (string fullname, AssemblyDefinition assembly, XPathNavigator nav)
+		{
+			Regex regex = new Regex (fullname.Replace (".", @"\.").Replace ("*", "(.*)"));
+
+			foreach (TypeDefinition type in assembly.MainModule.Types) {
+				MatchType (type, regex, nav);
+			}
+
+			if (assembly.MainModule.HasExportedTypes) {
+				foreach (var exported in assembly.MainModule.ExportedTypes) {
+					if (regex.Match (exported.FullName).Success) {
+						var type = ProcessExportedType (exported, assembly);
+						if (type != null) {
+							ProcessType (type, nav);
+						}
+					}
+				}
+			}
+
+			return true;
+		}
+
+		protected abstract void ProcessType (TypeDefinition type, XPathNavigator nav);
+
+		protected void ProcessTypeChildren (TypeDefinition type, XPathNavigator nav, object customData = null)
+		{
+			if (nav.HasChildren) {
+				ProcessSelectedFields (nav, type);
+				ProcessSelectedMethods (nav, type, customData);
+				ProcessSelectedEvents (nav, type, customData);
+				ProcessSelectedProperties (nav, type, customData);
+			}
+		}
+
+		void ProcessSelectedFields (XPathNavigator nav, TypeDefinition type)
+		{
+			XPathNodeIterator fields = nav.SelectChildren (FieldElementName, XmlNamespace);
+			if (fields.Count == 0)
+				return;
+
+			ProcessFields (type, fields);
```

Comment:
I thought we didn't want to have a bunch of small functions like ProcessSelected* calling Process* and instead, we will have the implementation directly on the code since the call doesn't add any value.

Reply:
I'll look into cutting it down a little bit. I went with this just because I copied most of the code from `ResolveFromXmlStep`. That said, quite a few of these methods are `protected virtual` on `ResolveFromXmlStep`, so removing them is effectively a breaking change.

Reply:
Thanks for the suggestion - relatively simple change and the file is now 50 lines shorter 👍